### PR TITLE
obs-studio: symlink libcef instead of copy

### PR DESCRIPTION
**Pull request by materusPL**
## Description of changes
Use symlink instead of copy for libcef. For issue https://github.com/NixOS/nixpkgs/issues/259165
Reduces space used by obs.

```bash
#before symlink
du -sh /nix/store/4q8l345f1d9h2bx81lzjrj15w0rr8147-obs-studio-29.1.3
426M    /nix/store/4q8l345f1d9h2bx81lzjrj15w0rr8147-obs-studio-29.1.3

#after symlink
du -sh /nix/store/bsqq89vrif4xb8gm2fxprq200iqd0md0-obs-studio-29.1.3
45M     /nix/store/bsqq89vrif4xb8gm2fxprq200iqd0md0-obs-studio-29.1.3
```
## Things done
Changed copy to symlink in post unpack phase.
Small change to pre and post fixup phase, remove and re link libcef so patchelfing will not fail. (Is there a better way like dontPatchELF for single path/lib?)

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by deviant** [comment]
I'm wondering if we can take this opportunity to make the code even simpler— the directory structure being created is the one that libcef originally comes in. Seems a little bit redundant for the libcef package expression to be rearranging its files, only for us to put them back where it found them! Perhaps splitting up the libcef package into two derivations, such that there's an intermediate step where all the files remain unmoved. Probably worth noting that (at least within Nixpkgs) obs-studio is the only consumer of the libcef package, so I'm not entirely sure who or what is benefiting from the rearranged version.

Even if we don't end up touching the libcef package here, I'd still suggest turning the CEF root into another derivation— this should resolve your patchelf woes, since it won't be touched by the hook.

**Comment by materusPL** [comment]
About libcef directory structure I guess it's question for maintainer @puffnfresh why it's structured like this/ if it can be changed. From my point of view it should be changed if it's used only by obs.

Question about how turning the CEF root into another derivation is supposed to help with patchelfing? obs seems to need libcef during buildtime if browser support is enabled (libcef.so too from what I checked) and it's copying it to output later.

**Comment by deviant** [comment]
> Question about how turning the CEF root into another derivation is supposed to help with patchelfing? obs seems to need libcef during buildtime if browser support is enabled (libcef.so too from what I checked)

I meant to create a `runCommand` that does the symlinking straight into its `$out`, then to pass that to the `-DCEF_ROOT_DIR=` option. Then it'll be available at build-time but won't get touched by the `autoPatchelfHook`. However, this might not even be necessary, because of:

> and it's copying it to output later.

I'd missed this one. We could patch the CMakeLists to symlink into the output instead— that way, we completely avoid vendoring anything, either at build-time or runtime. I think [here](https://github.com/obsproject/obs-studio/blob/5dde70ccfffd67e6115ae6a4a02e006ad91f94f7/cmake/Modules/ObsHelpers.cmake#L200-L223) is the place to change.

File: pkgs/applications/video/obs-studio/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/video/obs-studio/default.nix:149]:
```suggestion
    # remove libcef before patchelf, otherwise it will fail
    rm $out/lib/obs-plugins/libcef.so
```

File: pkgs/applications/video/obs-studio/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/video/obs-studio/default.nix:162]:
```suggestion
    # Link libcef again after patchelf for other libs
    ln -s ${libcef}/lib/libcef.so $out/lib/obs-plugins/libcef.so
```

**Comment by materusPL** [comment]
@deviant In my NUR I've made [patch](https://github.com/materusPL/Nixerus/blob/testing/pkgs/apps/obs/link-cef-instead-install.patch) to symlink cef instead of install in cmake.  Also changed [libcef](https://github.com/materusPL/Nixerus/blob/testing/pkgs/libs/libcef.nix) to have non changed hierarchy in `$out/share/cef`. That's my [obs](https://github.com/materusPL/Nixerus/blob/testing/pkgs/apps/obs/default.nix).

I still need to remove symlinks before patchelf hook , it seems to me patchelf hook tries to patch all .so  files, which will fail since it's symlink to read only nix store. Error if I won't rm symlinks before patchelf
```
stripping (with command strip and flags -S -p) in  /nix/store/qfidkfs3k6kvpz0jf1qx1f8lhlkh4q6n-obs-studio-amf-29.1.3/lib /nix/store/qfidkfs3k6kvpz0jf1qx1f8lhlkh4q6n-obs-studio-amf-29.1.3/bin
patchelf: open: Permission denied
```

Or did I misunderstood something about cmake?

**Comment by deviant** [comment]
> I still need to remove symlinks before patchelf hook , it seems to me patchelf hook tries to patch all .so files, which will fail since it's symlink to read only nix store. Error if I won't rm symlinks before patchelf
> 
> ```
> stripping (with command strip and flags -S -p) in  /nix/store/qfidkfs3k6kvpz0jf1qx1f8lhlkh4q6n-obs-studio-amf-29.1.3/lib /nix/store/qfidkfs3k6kvpz0jf1qx1f8lhlkh4q6n-obs-studio-amf-29.1.3/bin
> patchelf: open: Permission denied
> ```

That's unfortunate! I hadn't anticipated this would be a problem. In that case, I don't think we have a choice but to do that dance, unless `autoPatchelfHook` were to be modified to skip symlinks. Either way, I don't think that needs to hold this up from getting merged.

**Review by deviant — APPROVED** [review]

**Pull request merged**
